Convert to type script

russmiles opened the pull request.

Commit:
Added TypeScript support

Contributes to #11

Commit:
Added new typescript generator baseline

Contributes to #11

Commit:
Moved over to using TypeScript for this generator

Contributes to #11

ddgenome commented on .atomist/manifest.yml:
```diff
@@ -1,6 +1,6 @@
 group: atomist-rugs
 artifact: spring-boot-rest-service
-version: "0.7.1"
+version: "0.7.2"
```
Probably should be 0.8.0

ddgenome commented on .atomist.yml:
```diff
@@ -59,3 +59,32 @@ executor:
+    branch: "1979f9f98cad0eaf81eec3f512a80fdf3ae7d47f"
+    sha: "1979f9f"
+  parameters:
+    - "generator_name": "NewSpringBootRestServiceProj"
```
Was this editor run with the wrong generator name?

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+import { File } from '@atomist/rug/model/File'
+
+@Generator("NewSpringBootRestService", "create a new Spring Boot Rest Service project")
+@Tags("documentation", "java", "spring", "spring-boot", "spring-rest") 
```
Should probably remove the "documentation" tag.

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+    @Parameter({
+        displayName: "Maven Artifact ID",
+        description: "Maven artifact identifier, i.e., the name of the jar without the version, it is often the same as the project name",
+        pattern: "^[a-z][-a-z0-9_]*$", // Ideally this should be looking up artifact_id as a common pattern
```
Isn't there a set pattern for this? Oh, perhaps this was the one that was missing that you created an issue for?

russmiles replied:
Is indeed the one that was missing, for now.

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+    @Parameter({
+        displayName: "Group ID",
+        description: "Maven group identifier, often used to provide a namespace for your project, e.g., com.pany.team",
+        pattern: "^[a-z][-a-z0-9_]*$", // Ideally this should be looking up artifact_id as a common pattern
```
I'm pretty sure there is a `Pattern.group_id` regex.

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+        pattern: Pattern.semantic_version,
+        validInput: "a valid semantic version, http://semver.org",
+        minLength: 1,
+        maxLength: 21,
```
Is 21 long enough?

russmiles replied:
Possibly not, lengthening to 50

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+        pattern: Pattern.any,
+        validInput: "free text",
+        minLength: 1,
+        maxLength: 50,
```
Typically we limit this to 100

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+        pattern: Pattern.java_package,
+        validInput: "a valid Java package name, which consists of period-separated identifiers which have only alphanumeric characters, $ and _ and do not start with a number",
+        minLength: 1,
+        maxLength: 50,
```
Is 50 long enough?

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    switchREADME(project) {
+        project.merge("readme.vm", "README.md", {"project_name": this.project_name, "description": this.description});
```
Perhaps a good time to move away from templates too?

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+        this.tidyProject(project);
+    }
+
+    switchREADME(project) {
```
Shouldn't all of these function signatures provide a type for `project`, i.e., `project: Project`?

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,170 @@
+        });
+    }
+
+    packageMove(project, new_package) {
```
`new_project` is a `string`? TypeScript standards requires camelCase for function and variable names. So `new_package` should be `newPackage`.

Similar for all variables below.

ddgenome reviewed (changes requested):
Shouldn't the `NewSpringBootRestService.rug` file have been removed rather than just moved to the top of the project?

Commit:
Incorporated review comments

Contributes to #11

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -131,33 +140,37 @@ class NewSpringBootRestService implements PopulateProject {
+        readme.replace("{{description}}", this.description);
 
-        readmeFile.replace("{{creation_date}}", new Date().toISOString().split('T')[0]);
+        readme.replace("{{creation_date}}", new Date().toISOString().split('T')[0]);
```
If you're going to do it this way, you might as well keep using the template.

Here's an example that uses the README from the generator's archive, replacing and excising as appropriate.

ddgenome reviewed (commented):
Truly do not use templates.

Commit:
Removed button links

Contributes to #11

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,183 @@
+        this.tidyProject(project);
+    }
+
+    selectReadmeFile(project:Project) {
```
Shouldn't this also specify a return type?

```typescript
selectReadmeFile(project: Project): File {
```

ddgenome commented on .atomist/editors/NewSpringBootRestService.ts:
```diff
@@ -0,0 +1,183 @@
+        readme.replace("{{description}}", this.description);
+
+        readme.replace("{{creation_date}}", new Date().toISOString().split('T')[0]);
+    }
```
If you're going to handle the README this way, you might as well keep using the template.

Here's an example that uses the README from the generator's archive, replacing and excising as appropriate.

russmiles replied:
See your point, but don't see the example?

I'll move to pure README amendments instead, it makes sense.

ddgenome replied:
Sorry copy/paste error. Here's the link to the example:
https://github.com/atomist-rugs/rug-project/blob/master/.atomist/editors/RugGeneratorFunctions.ts#L34-L44

russmiles replied:
Thanks! I'll incorporate that approach right now.

ddgenome reviewed (commented):
A few more questions

Commit:
Added functions to replace contents of README and CHANGELOG

Contributes to #11

ddgenome commented on README.md:
```diff
@@ -3,12 +3,7 @@
-[rug]: http://docs.atomist.com/
-[boot]: https://projects.spring.io/spring-boot/
-[rest]: https://en.wikipedia.org/wiki/Representational_state_transfer
+This Rug archive contains a Generator for a Spring Boot Rest Service project.
```
We prefer lowercase generator.

ddgenome commented on .atomist/manifest.yml:
```diff
@@ -1,6 +1,6 @@
-version: "0.7.1"
+version: "0.8.0"
 
 requires: "[0.10.0,1.0.0)"
```
Update to 0.12.0?

ddgenome reviewed (commented):
Perhaps I should have made those functions editors so you could have called them instead of copying them.

A couple minor things and then it is good to merge.

Commit:
Lowercase generator in README and moved to Rug 0.12.0

Contributes to #11

russmiles commented:
Like the idea of making those functions editors in the future. For now just those functions locally and if we decide that's the way to go it can be a new issue here and in the common-editors @ddgenome ?

ddgenome commented:
Are you going to change the rug dependency to 0.12.0?

Why did you remove the links in the README and leave generator upper case?

russmiles commented:
The links were removed as they made the regex heavily confused for replacement. Happy to reintroduce if I can get your help on making the regex replace work?

Good point on the CHANGELOG for sure, I'll get that entry done now and push.

russmiles commented:
Commits will be squashed on "Squash and Merge"

Commit:
Updated CHANGELOG

Contributes to #11

ddgenome commented on .atomist/editors/RugGeneratorFunctions.ts:
```diff
@@ -0,0 +1,47 @@
+    let readMePE = new PathExpression<Project, File>("/*[@name='README.md']");
+    let readMe: File = eng.scalar(project, readMePE);
+    readMe.replace("# Atomist 'spring-boot-rest-service'", "# " + projectName);
+    readMe.regexpReplace("This Rug archive contains a generator for a Spring Boot Rest Service project.[\\s\\S]*?\n## Rugs\n", description + "\n");
```
If you have the following text in the README:

```markdown
This [Rug][rug] project contains a generator for a [Spring Boot][boot]
[REST][rest] service project.

[rug]: http://docs.atomist.com/
[boot]: https://projects.spring.io/spring-boot/
[rest]: https://en.wikipedia.org/wiki/Representational_state_transfer

## Rugs
```

just alter the regex to skip the bad parts. Something like,

```typescript
readMe.regexpReplace("This .*?Rug.*? project contains a generator for a .*?Spring Boot[\\s\\S]*?\n## Rugs\n", description + "\n\n##Rugs\n");
```

russmiles replied:
Nice! Thanks @ddgenome ; I'll add those pieces in now.

ddgenome reviewed (commented):
I added a comment on the text and regex.

Commit:
Added links back into README

Contributes to #11

ddgenome commented on .atomist/editors/RugGeneratorFunctions.ts:
```diff
@@ -0,0 +1,47 @@
+    let readMePE = new PathExpression<Project, File>("/*[@name='README.md']");
+    let readMe: File = eng.scalar(project, readMePE);
+    readMe.replace("# Atomist 'spring-boot-rest-service'", "# " + projectName);
+    readMe.regexpReplace("This .*?Rug.*? project contains a generator for a .*?Spring Boot[\\s\\S]*?\n## Rugs\n", description + "\n");
```
Seems you missed part of the replacement text. Shouldn't it be `"\n\n##Rugs\n"`?

russmiles replied:
I deliberately removed that as it produces an odd `## Rugs` header in the resulting README with no content as the generated project is not a rug archive.

ddgenome replied:
Yes, of course. We should probably simplify the code a bit then, eh? We really only need one replaceRegexp.

ddgenome reviewed (commented):
One more thing

ddgenome commented on .atomist/editors/RugGeneratorFunctions.ts:
```diff
@@ -0,0 +1,47 @@
+    let readMe: File = eng.scalar(project, readMePE);
+    readMe.replace("# Atomist 'spring-boot-rest-service'", "# " + projectName);
+    readMe.regexpReplace("This .*?Rug.*? project contains a generator for a .*?Spring Boot[\\s\\S]*?\n## Rugs\n", description + "\n");
+    readMe.regexpReplace("\n### NewSpringBootRestService[\\s\\S]*\n## Support\n", "\n## Support\n");
```
The above two `regexpReplace` can be replaced with a single:

```typescript
readMe.regexpReplace("This .*?Rug.*? project contains a generator for a .*?Spring Boot[\\s\\S]*?\n## Support\n", description + "\n\n## Support\n");
```

ddgenome commented on CHANGELOG.md:
```diff
@@ -7,7 +7,17 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
+
+## [0.7.1] - 2017-01-03
+
+[0.7.0]: https://github.com/atomist-rugs/spring-boot-rest-service/compare/0.6.0...0.7.1
```
What version are we releasing? These numbers are all over the place. Is 0.6.0 that last release? If so, this should be 0.7.0.

russmiles replied:
0.7.1 is our last release that I can see?

ddgenome reviewed (commented):
Combine statements and question about release version.

Commit:
Corrected CHANGELOG.md and simplified regex replace for resulting README

Contributes to #11

Commit:
Added simple Gherkin test

Commit:
New tests

cdupuis committed:
Make tests pass

Commit:
Bring generator up to current standards

Clean up existing TypeScript.  Convert DSL tests to TypeScript Gherkin
tests using TestsToTypeScript and then clean up and consolidate.  Add
more tests to ensure proper behavior.  Remove dependency on
common-editors.  Use camelCase parameters and variables.  Update to
latest version of rug.  Add Spring Boot/Maven instructions to README.
Clean up POM.
